src/operator/installer.ts

```diff
@@ -0,0 +1,130 @@
+import { terminal } from '../common/debug'
+import { createUpdateConfigMap, createUpdateGenericSecret, getK8sConfigMap, k8s } from '../common/k8s'
+import { hfValues } from '../common/hf'
+import { AplOperations } from './apl-operations'
+import { getErrorMessage } from './utils'
+import * as process from 'node:process'
+
+export interface GitCredentials {
+  username: string
+  password: string
+}
+
+export class Installer {
+  private d = terminal('operator:apl-installer')
+  private aplOps: AplOperations
+
+  constructor(aplOps: AplOperations) {
+    this.aplOps = aplOps
+    this.d.info('Initializing APL Installer')
+  }
+
+  public async reconcileInstall(): Promise<void> {
+    let attemptNumber = 0
+
+    while (true) {
+      attemptNumber += 1
+      this.d.info(`Starting installation attempt ${attemptNumber}`)
+
+      try {
+        // Always run bootstrap to ensure the environment is ready (even on restarts)
+        await this.aplOps.validateCluster()
+        await this.aplOps.bootstrap()
+
+        // Check if installation already completed
+        const installStatus = await this.getInstallationStatus()
+        if (installStatus === 'completed') {
+          this.d.info('Installation already completed, skipping install steps')
+          return
+        }
+
+        // Run the installation sequence
+        await this.updateInstallationStatus('in-progress', attemptNumber)
+        await this.aplOps.install()
+
+        await this.updateInstallationStatus('completed', attemptNumber)
+        this.d.info('Installation completed successfully')
+        return
+      } catch (error) {
+        const errorMessage = getErrorMessage(error)
+        this.d.error(`Installation attempt ${attemptNumber} failed:`, errorMessage)
+        await this.updateInstallationStatus('failed', attemptNumber, errorMessage)
+        this.d.warn(`Installation attempt ${attemptNumber} failed, retrying in 1 second...`, getErrorMessage(error))
+
+        // Wait 1 second before retrying
+        await new Promise((resolve) => setTimeout(resolve, 1000))
+      }
+    }
+  }
+
+  private async getInstallationStatus(): Promise<string> {
+    try {
+      const configMap = await getK8sConfigMap('apl-operator', 'apl-installation-status', k8s.core())
+      const status = configMap?.data?.status || 'pending'
+      this.d.info(`Current installation status: ${status}`)
+      if (configMap?.data) {
+        this.d.info(`ConfigMap data:`, configMap.data)
```

Comment:
It currently is echoed to the logs, but seems unnecessarily verbose. I would suggest
```suggestion
        this.d.debug(`ConfigMap data:`, configMap.data)
```
also I would move this above line 63 and also let it show up on an empty response, so it is more clear.